lib/net.js

```diff
@@ -551,13 +551,9 @@ function onReadableStreamEnd() {
 
 
 Socket.prototype.destroySoon = function() {
+  stream.finished(this, { error: false, readable: false }, this.destroy);
```

Comment:
Note sure if the callback can be called synchronously? Just to be safe... put it after the `end()` call?

Reply:
`eos()` unconditionally adds one listener for the `'finish'` event and one for `'close'` event (and actually another for `'end'` event even if the `readable` option is `false`).

If `_writableState.finished` is already `true` then the `'finish'` event won't be emitted and it will wait for `'close'`. Even if this is emitted synchronously it is not a problem because it means the socket is already destroyed and no longer writable.

If `_writableState.finished` is `false` then the `'finish'` event won't be emitted until `socket.end()` is called and if this already happened then the socket is no longer writable.

So the order should not matter. However I can see a case where the behavior is different.

```js
'use strict';

const net = require('net');

const server = net.createServer(function(socket) {
  socket.write('foo');
});

server.listen(function() {
  const socket = net.connect(this.address().port, function() {
    socket.on('finish', function() {
      socket.destroySoon();
    });
    socket.on('close', function() {
      console.log('close');
    });

    setTimeout(function() {
      socket.end();
    }, 100);
  });
});
```

Yes, it's a fabricated example that probably will never happen in reality but with this patch the socket is not destroyed (`'close'` is not emitted) until the buffered data is actually read. This is not the case with the current implementation of `destroySoon()`.